openwisp_firmware_upgrader/api/views.py

```diff
@@ -0,0 +1,117 @@
+from rest_framework import generics
+from rest_framework.permissions import IsAuthenticated
+from swapper import load_model
+
+from openwisp_users.api.authentication import BearerAuthentication
+
+from .serializers import (
+    BatchUpgradeOperationSerializer,
+    BuildSerializer,
+    CategorySerializer,
+    FirmwareImageSerializer,
+)
+
+BatchUpgradeOperation = load_model('firmware_upgrader', 'BatchUpgradeOperation')
+Build = load_model('firmware_upgrader', 'Build')
+Category = load_model('firmware_upgrader', 'Category')
+FirmwareImage = load_model('firmware_upgrader', 'FirmwareImage')
+
+
+class ProtectedAPIMixin(object):
+    authentication_classes = [BearerAuthentication]
+    permission_classes = [IsAuthenticated]
+
+
+class OrgAPIMixin(object):
+    def get_queryset(self):
+        queryset = self.queryset.all()
+        if not self.request.user.is_superuser:
+            user_org = self.request.user.openwisp_users_organization.get()
+            organization_filter = {self.organization_field: user_org}
+            queryset = queryset.filter(**organization_filter)
+        org = self.request.query_params.get('org', None)
+        try:
+            if org:
+                organization_filter = {self.organization_field + '__name': org}
+                queryset = queryset.filter(**organization_filter)
+        except Exception:
+            # django.core.exceptions.ValidationError when uuid is not valid
+            queryset = []
+        return queryset
+
+
+class BuildListView(ProtectedAPIMixin, OrgAPIMixin, generics.ListCreateAPIView):
+    queryset = Build.objects.all()
+    serializer_class = BuildSerializer
+    organization_field = 'category__organization'
+
+
+class BuildDetailView(ProtectedAPIMixin, generics.RetrieveUpdateDestroyAPIView):
+    queryset = Build.objects.all()
+    serializer_class = BuildSerializer
+    lookup_fields = ['pk']
+    permission_classes = [IsAuthenticated]
```

Comment:
what happens if somebody tries to access a build which is related to an organization the user does not have access to?

Reply:
Well spotted. I've fixed it and now returns 404 if the user does not belong to the organization